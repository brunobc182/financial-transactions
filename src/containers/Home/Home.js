import React, { Component } from 'react';
import styled from 'styled-components';
import { Button, Input, TransactionList } from '../../components';

export default class Home extends Component {
  constructor() {
    super();
    this.state = {
      amount: 0,
      transactions: [],
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleNewTransaction = this.handleNewTransaction.bind(this);
  }

  handleChange(event, maskedvalue, floatvalue) {
    this.setState({
      amount: floatvalue,
    });
  }

  handleNewTransaction() {
    const { amount } = this.state;
    this.setState(prevState => ({
      transactions: [...prevState.transactions, { type: 'DEBIT', value: amount }],
    }));
  }

  render() {
    const { amount, transactions } = this.state;
    return (
      <Wrapper>
        <Container>
          <Input value={amount} onChange={this.handleChange} />
          <ButtonWrapper>
            <Button onClick={this.handleNewTransaction}>+ Credit</Button>
            <Button onClick={this.handleNewTransaction} isDebit>
              - Debit
            </Button>
          </ButtonWrapper>
        </Container>
        <ListWrapper>
          <TransactionList data={transactions} />
        </ListWrapper>
      </Wrapper>
    );
  }
}

const Wrapper = styled.section`
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  max-width: 320px;
  width: 100%;
  margin-top: 30px;
`;
const ButtonWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  width: 100%;
  margin-top: 10px;
`;

const ListWrapper = styled.div`
  max-width: 460px;
  width: 100%;
`;
