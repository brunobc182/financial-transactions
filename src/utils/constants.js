const TRANSACTION_TYPE = {
  DEBIT: 'DEBIT',
  CREDIT: 'CREDIT',
};

export default TRANSACTION_TYPE;
