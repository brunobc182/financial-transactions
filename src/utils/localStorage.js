const setItem = (name, item) => localStorage.setItem(name, JSON.stringify(item));
const getItem = name => JSON.parse(localStorage.getItem(name));

export { setItem, getItem };
