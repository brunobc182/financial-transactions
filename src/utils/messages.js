const MESSAGES = {
  totalLabel: 'Total: ',
  partialTotalLabel: 'Partial total: ',
};

export default MESSAGES;
