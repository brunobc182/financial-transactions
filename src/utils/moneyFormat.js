const moneyFormat = (number, prefix = '$') => `${prefix} ${number
  .toFixed(2)
  .replace(/(\d)(?=(\d{3})+\.)/g, '$1.')
  .replace(/\.(\d{2})$/, ',$1')}`;

export default moneyFormat;
