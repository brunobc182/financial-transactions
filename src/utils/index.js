import { TRANSACTION_TYPE, LOCAL_STORAGE_NAME } from './constants';
import moneyFormat from './moneyFormat';
import MESSAGES from './messages';
import { getItem, setItem } from './localStorage';

export {
  LOCAL_STORAGE_NAME, TRANSACTION_TYPE, moneyFormat, MESSAGES, getItem, setItem,
};
