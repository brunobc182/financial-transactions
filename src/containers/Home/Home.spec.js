import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Home from './Home';
import { TRANSACTION_TYPE } from '../../utils';

jest.mock('react-currency-input');

describe('Home', () => {
  const { CREDIT, DEBIT } = TRANSACTION_TYPE;

  const initialState = {
    amount: 0,
    transactions: [],
  };

  it('should render correctly', () => {
    const tree = renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should change the amount value', () => {
    const tree = mount(<Home />);
    const value = 1.12;

    tree.setState({ ...initialState });
    tree
      .find('Home')
      .instance()
      .handleChange(null, null, value);

    expect(tree.state().amount).toBe(1.12);
  });

  it('should delete a transaction', () => {
    const tree = mount(<Home />);
    const index = 0;
    const newTransactions = [
      {
        type: DEBIT,
        value: -1,
        partialTotal: 9,
      },
    ];

    tree.setState({
      ...initialState,
      transactions: [
        {
          type: CREDIT,
          value: 10,
          partialTotal: 10,
        },
        {
          type: DEBIT,
          value: -1,
          partialTotal: 9,
        },
      ],
    });

    tree
      .find('Home')
      .instance()
      .handleDeleteTransaction(index);

    expect(tree.state().transactions).toEqual(newTransactions);
  });

  it('should add a new CREDIT transaction', () => {
    const tree = mount(<Home />);
    const newTransactions = [
      {
        type: CREDIT,
        value: 10,
        partialTotal: 10,
      },
    ];

    tree.setState({
      ...initialState,
      amount: 10,
    });

    tree
      .find('Home')
      .instance()
      .handleNewTransaction(CREDIT);

    expect(tree.state().transactions).toEqual(newTransactions);
  });

  it('should add a new DEBIT transaction', () => {
    const tree = mount(<Home />);
    const newTransactions = [
      {
        type: DEBIT,
        value: -90,
        partialTotal: -90,
      },
    ];

    tree.setState({
      ...initialState,
      amount: 90,
    });

    tree
      .find('Home')
      .instance()
      .handleNewTransaction(DEBIT);

    expect(tree.state().transactions).toEqual(newTransactions);
  });
});
