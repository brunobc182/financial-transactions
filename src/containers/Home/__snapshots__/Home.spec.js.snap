// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<section
  className="sc-ifAKCX bMIgkK"
>
  <div
    className="sc-EHOje jzrzrp"
  >
    <div
      className="sc-bZQynM hjHjiY"
    >
      <button
        className="sc-bdVaJa aajhJ"
        onClick={[Function]}
        type="button"
      >
        + Credit
      </button>
      <button
        className="sc-bdVaJa bEifWJ"
        onClick={[Function]}
        type="button"
      >
        - Debit
      </button>
    </div>
  </div>
  <div
    className="sc-gzVnrw hkMjPv"
  >
    <ul
      className="sc-htpNat fyjIhF"
    />
  </div>
</section>
`;
